### ISSUE TITLE
Restart deployment/daemonset when referred resources' content is changing

### ISSUE THREAD

**Issue by chenliu1993** [k14s/kapp#286]
Hi team, I met a senario when using kapp's versioned annotation.  Opened this issue to track the question.
The scenario is:
**config C refers a secret S in its data,  and this config C is referred by a Deployment D. Will update the content of secret S trigger restart of Deployment D?**

I get a solution that is referring the secret S directly in Depoyment D. But I think that may be not the best way.

Will kapp's versioned annotation supports above scenario? Or already supported? 
Thanks.

**Comment by cppforlife** [comment]
> config C refers a secret S in its data, and this config C is referred by a Deployment D. Will update the content of secret S trigger restart of Deployment D?

it will not today, since kapp does not support updating resource references within arbitrary content (data of ConfigMap). im not sure we want to extend kapp to try to update contents of ConfigMaps with new resource name. (we do something like this in kbld to update image refs but it only works for json/yaml content).

> I get a solution that is referring the secret S directly in Depoyment D. But I think that may be not the best way.

currently, that's the only way to make this work with versioned resources.  note that referencing a secret on the deployment does not mean that you Deployment needs to actually consume it directly, it could still be doing it through changed ConfigMap.

outside of versioned resources feature, you could do what folks do with templating tricks to hash content and throw in an annotation onto Deployment, but to me that feels more work that referencing secret on the Deployment.

as an enhancement to versioned resources feature, i could imagine supporting special annotation that will make kapp consider ConfigMap as related to changed Secret which then would trigger Deployment change. e.g.

```yaml
kind: Secret
metadata:
  name: something
  annotations:
   kapp.k14s.io/versioned: ""
# ...
---
kind: ConfigMap
metadata:
  name: config
  annotations:
   kapp.k14s.io/versioned: ""
   kapp.k14s.io/versioned-explicit-ref: "v1/Secret/something" <--- this would be "synthetic reference"
# ...
---
kind: Deployment # only references ConfigMap
# ...
```

**Comment by chenliu1993** [comment]
Thanks @cppforlife . The information is really detailed😊 and I would vote for the example solution `kapp.k14s.io/versioned-explicit-ref` if it is doable.

**Comment by renuy** [comment]
Moving it to carvel accepted. Will try and prioritise this

**Comment by renuy** [comment]
Pulling this issue for the next release v0.41.0 . @100mik , please update the approach here.

**Comment by 100mik** [comment]
We can support a synthetic reference which mirrors the latest versioned name of a versioned resource.

For example, a synthetic reference like
```yaml
kapp.k14s.io/versioned-explicit-ref: default/v1/Secret/credentials
```
will be assigned the following value on deployment
```yaml
kapp.k14s.io/versioned-explicit-ref: credentials-ver-1
```
This leads to the referencing versioned resource being versioned whenever the referenced resource is. It can be used to explicitly refer a versioned resource from a non-versioned resource as well.

Ongoing work on this with [this draft PR](https://github.com/vmware-tanzu/carvel-kapp/pull/311)!

**Comment by 100mik** [comment]
Addressing the questions raised in the ongoing [PR](https://github.com/vmware-tanzu/carvel-kapp/pull/311), we will allow users to reference a single or multiple resources in a JSON object with specified format.

```yaml
kapp.k14s.io/versioned-explicit-ref: '{ "references": [ { "namespace": <resource-namespace>, "apiGroup":  <resource-apiGroup>, "kind" : <resource-kind>, "name":  <resource-name> } ] }'
``` 
- `references` can have multiple objects referring to different resources
- The `namespace` key should be omitted for cluster scoped resources

@chenliu1993 I believe this will satisfy your requirements ?

**Comment by 100mik** [comment]
As of v0.41.0+, versioned resources can be explicitly referred to by the `kapp.k14s.io/versioned-explicit-ref` annotations like
```yaml
kapp.k14s.io/versioned-explicit-ref: |
      apiVersion: v1
      kind: ConfigMap
      name: config-1 
```
(a `namespace` key can be added for namespaces resources). Documentation regarding the same canoe found at the end of [this](https://carvel.dev/kapp/docs/latest/diff/#versioned-resources) section in our documentation.

**Comment by renuy** [comment]
Hi @chenliu1993, This feature is now available as part of v0.41.0 . Do provide feedback on this.
Thanks
